server/src/server.js

```diff
@@ -186,6 +196,132 @@ app.get("/:id/:domain", function (req, res) {
+  profileBaseURI = 'http://127.0.0.1:1111/v1',
+  oAuthClientId = 'ac7ee3c317531aab',
+  oAuthClientSecret = 'eb7c92bd4616a7c1f86595492f360f55cc453974b062e71456d2d4b104f307f8';
+
+// Get OAuth client params for the client-side authorization flow.
+app.get('/api/fxa-oauth/params', function (req, res, next) {
+  if (! req.userId) {
+    next(errors.sessionRequired());
+    return;
+  }
+  helpers.randomBytes(32).then(state => {
+    return setState(req.userId, state).then(inserted => {
+      if (!inserted) {
+        throw errors.dupeLogin();
+      }
+      let stateHex = state.toString('hex');
+      return stateHex;
+    });
+  }).then(stateHex => {
+    res.send({
+      // FxA OAuth server URL.
+      oauth_uri: oAuthBaseURI,
+      redirect_uri: 'urn:ietf:wg:oauth:2.0:fx:webchannel',
+      client_id: oAuthClientId,
+      // FxA content server URL.
+      content_uri: contentBaseURI,
+      state: stateHex,
+      scope: 'profile'
+    });
+  }).catch(next);
+});
+
+// Exchange an OAuth authorization code for an access token.
+app.post('/api/fxa-oauth/token', function (req, res, next) {
+  if (! req.userId) {
+    next(errors.sessionRequired());
+    return;
+  }
+  if (! req.body) {
+    next(errors.paramsRequired());
+    return;
+  }
+  let { code, state: stateHex } = req.body;
+  let state = new Buffer(stateHex, 'hex');
+  checkState(req.userId, state).then(isValid => {
+    if (!isValid) {
+      throw errors.invalidState();
+    }
+    let oAuthURI = `${oAuthBaseURI}/token`;
+    return helpers.request('POST', oAuthURI, {
+      payload: JSON.stringify({
+        code,
+        client_id: oAuthClientId,
+        client_secret: oAuthClientSecret
+      }),
+      headers: {
+        'content-type': 'application/json'
+      },
+      json: true
+    }).then(([profileRes, body]) => {
+      if (profileRes.statusCode < 200 || profileRes.statusCode > 299) {
+        throw errors.badToken();
+      }
+      let { access_token: accessTokenHex } = body;
+      let accessToken = new Buffer(accessTokenHex, 'hex');
+      return getProfile(accessTokenHex).then(profile => {
+        let {
+          uid: accountIdHex,
+          email
+        } = profile;
+        let accountId = new Buffer(accountIdHex, 'hex');
+        return db.transaction(client => {
+          return db.upsertWithClient(
+            client,
+            `INSERT INTO accounts (id, token, email) SELECT $1, $2, $3`,
+            `UPDATE accounts SET token = $2, email = $3 WHERE id = $1`,
+            [accountId, accessToken, email]
+          ).then(() => {
+            return db.queryWithClient(
+              client,
+              `UPDATE devices SET accountid = $2 WHERE id = $1`,
+              [accountId, req.userId]
+            );
+          });
+        }).then(() => {
+          res.send({
+            access_token: accessTokenHex
+          });
+        });
+      });
+    });
+  }).catch(next);
+});
+
+function setState(deviceId, state) {
```

Comment:
I'd like to keep SQL itself out of server.js.  Probably a new module to handle these?